File: Android/app/src/main/java/catglo/com/deliverydroid/settings/SettingsActivity.java
package catglo.com.deliverydroid.settings;

import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;
import android.content.SharedPreferences.OnSharedPreferenceChangeListener;
import android.location.Criteria;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Bundle;
import android.os.Environment;
import android.preference.CheckBoxPreference;
import android.preference.EditTextPreference;
import android.preference.Preference;
import android.preference.Preference.OnPreferenceChangeListener;
import android.preference.Preference.OnPreferenceClickListener;
import android.preference.PreferenceActivity;
import android.preference.PreferenceManager;
import android.preference.PreferenceScreen;
import android.widget.Toast;
import catglo.com.deliveryDatabase.DataBase;
import catglo.com.deliveryDatabase.StreetList;
import catglo.com.deliveryDatabase.ZipCode;
import catglo.com.deliverydroid.AboutActivity;
import catglo.com.deliverydroid.BuildConfig;
import catglo.com.deliverydroid.R;
import catglo.com.deliverydroid.backup.GoogleDriveBackupRestoreActivity;
import catglo.com.widgets.TwoLinesListPreference;


import java.io.File;
import java.io.FileInputStream;
import java.io.FileNotFoundException;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.util.Iterator;
import java.util.regex.Pattern;



public class SettingsActivity extends PreferenceActivity implements OnSharedPreferenceChangeListener {
	EditTextPreference			databaseFileCopy;
	CheckBoxPreference			databaseOnSdcard;

	SharedPreferences			prefs;
	private PreferenceScreen	zipCodePrefs;
	private EditTextPreference	newZipCode;
	Pattern						pattern;
	private StreetList streetList;
	DataBase dataBase;

	private Editor pedit;
	private EditTextPreference databaseFileMerge;
	private EditTextPreference databaseFileUse;
	private Preference rebuildStreetList;
	private EditTextPreference centrPoint_lat_s;
	private EditTextPreference centrPoint_lng_s;
	private Preference rebuildRegionsList;
	private TwoLinesListPreference navigationListPref;
    private Preference aboutPref;
    private Preference backupToGoogleDrive;

    @Override
	protected void onDestroy() {
		dataBase.close();
		super.onDestroy();
	}
	
	 @Override
	  public void onStart() {
	    super.onStart();

	   // EasyTracker.getInstance(this).activityStart(this);  // Add this method.
	  }

	  @Override
	  public void onStop() {
	    super.onStop();
	    
	   // EasyTracker.getInstance(this).activityStop(this);  // Add this method.
	  }
	
	/** Called when the activity is first created. */
	@SuppressWarnings("deprecation")
	@Override
	public void onCreate(final Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		
		addPreferencesFromResource(R.xml.settings);
	

	    if (dataBase == null) {
        	dataBase = new DataBase(getApplicationContext());
        	dataBase.open();
        }
	    
	    
	    navigationListPref = (TwoLinesListPreference)getPreferenceScreen().findPreference("navigationIntent");
	    navigationListPref.setEntriesSubtitles(getResources().getTextArray(R.array.navigation_options_summary));
	    
	    
//		InstructionsPreference instructionsAltPay = (InstructionsPreference) getPreferenceScreen().findPreference("instructionsAltPay");
//		instructionsAltPay.
		
		streetList = StreetList.LoadState(getApplicationContext());
		
		prefs = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
		pedit = prefs.edit();
		pedit.putString("centrPoint_lat_s", prefs.getFloat("centrPoint_lat", 0)+"");
		pedit.putString("centrPoint_lng_s", prefs.getFloat("centrPoint_lng", 0)+"");
		pedit.commit();
		
		centrPoint_lat_s = (EditTextPreference) getPreferenceScreen().findPreference("centrPoint_lat_s");
		centrPoint_lng_s = (EditTextPreference) getPreferenceScreen().findPreference("centrPoint_lng_s");
		centrPoint_lat_s.setOnPreferenceChangeListener(new OnPreferenceChangeListener(){

			public boolean onPreferenceChange(Preference preference,Object newValue) {
				pedit = prefs.edit();
				float value;
				try {
					value = Float.parseFloat(newValue.toString());
				} catch (NumberFormatException e){
					value = 100;
					Toast.makeText(getApplicationContext(), R.string.error_decoding_number, Toast.LENGTH_LONG).show();
				}
				pedit.putFloat("centrPoint_lat", value);
				pedit.putString("centrPoint_lat_s", newValue.toString());
				pedit.commit();
				return false;
			}});
		
		centrPoint_lng_s.setOnPreferenceChangeListener(new OnPreferenceChangeListener(){

			public boolean onPreferenceChange(Preference preference, Object newValue) {
				pedit = prefs.edit();
				float value;
				try {
					value = Float.parseFloat(newValue.toString());
				} catch (NumberFormatException e){
					value = 100;
					Toast.makeText(getApplicationContext(), R.string.error_decoding_number, Toast.LENGTH_LONG).show();
				}
			
				pedit.putFloat("centrPoint_lng", value );
				pedit.putString("centrPoint_lng_s", newValue.toString());
				pedit.commit();
				return false;
			}});
		//SettingsStoreAddress storeAddress = (SettingsStoreAddress) getPreferenceScreen().findPreference("storeAddress");
		
		
	    {
			databaseFileCopy = (EditTextPreference) getPreferenceScreen().findPreference("DatabaseFileCopy");
			databaseOnSdcard = (CheckBoxPreference) getPreferenceScreen().findPreference("DatabaseOnSdcard");
			//databaseUpload   = (CheckBoxPreference) getPreferenceScreen().findPreference("DatabaseUpload");
			databaseFileMerge = (EditTextPreference) getPreferenceScreen().findPreference("DatabaseFileMerge");
			databaseFileUse = (EditTextPreference) getPreferenceScreen().findPreference("DatabaseFileUse");
			rebuildStreetList = (Preference) getPreferenceScreen().findPreference("rebuildStreetList");
			rebuildRegionsList = (Preference)getPreferenceScreen().findPreference("rebuildRegionsList");
            aboutPref = (Preference)getPreferenceScreen().findPreference("aboutPref");
            backupToGoogleDrive = (Preference)getPreferenceScreen().findPreference("backupToGoogleDrive");

/*
			rebuildStreetList.setOnPreferenceClickListener(new OnPreferenceClickListener(){

				public boolean onPreferenceClick(Preference preference) {
					final Iterator<ZipCode> keys = StreetList.zipCodes.iterator();
					while (keys.hasNext()) {
						final ZipCode z = keys.next();
						z.state = ZipCode.STATE_NEEDS_LOOKUP;
					}
					final ProgressDialog dialog = ProgressDialog.show(SettingsActivity.this, "", 
	                        "Downloading. Please wait...", false);
					dialog.show();
					Thread t = new Thread(new Runnable(){public void run() {
						streetList.run();
						runOnUiThread(new Runnable(){public void run() {
							dialog.dismiss();
						}});
					}});
					t.start();
					
					return false;
				}

			});
			
			rebuildRegionsList.setOnPreferenceClickListener(new OnPreferenceClickListener(){public boolean onPreferenceClick(Preference arg0) {
				final ProgressDialog dialog = ProgressDialog.show(SettingsActivity.this, "", 
                        "Waiting for GPS lock. Please wait...", false);
				dialog.show();
				
				LocationManager locationManager = (LocationManager) getSystemService(Context.LOCATION_SERVICE);
				Criteria criteria = new Criteria();
				criteria.setAccuracy(Criteria.ACCURACY_LOW);
				locationManager.requestSingleUpdate(criteria, new LocationListener(){
					public void onLocationChanged(Location location) {
						
						final Editor prefEditor = prefs.edit();
						prefEditor.putFloat("centrPoint_lat", (float) location.getLatitude());
						prefEditor.putFloat("centrPoint_lng", (float) location.getLongitude());
						prefEditor.putString("centrPoint_lat_s", location.getLatitude()+"");
						prefEditor.putString("centrPoint_lng_s", location.getLongitude()+"");	
									
						dialog.setTitle("Downloading, Please Wait...");
						
						new WebServiceUpdateLocalePreferances(prefs,(float) location.getLatitude(),(float) location.getLongitude(),new WebServiceUpdateLocalePreferances.LocalityListener(){public void result(LocalityInfo info) {
							prefEditor.putString("addressFilterComponents", info.filter);
							prefEditor.commit();
							runOnUiThread(new Runnable(){public void run() {
								dialog.dismiss();
							}});
						}}).lookup();
						
					}
					public void onProviderDisabled(String arg0) {
						Toast.makeText(getApplicationContext(), "Error No GPS", Toast.LENGTH_LONG).show();						
					}
					public void onProviderEnabled(String arg0) {}
					public void onStatusChanged(String arg0, int arg1, Bundle arg2) {}
				}, null);
				
				return false;
			}});
*/
            aboutPref.setOnPreferenceClickListener(new OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    Intent i = new Intent(getApplicationContext(), AboutActivity.class);
                    startActivity(i);
                    return false;
                }
            });

            backupToGoogleDrive.setOnPreferenceClickListener(new OnPreferenceClickListener() {
                @Override
                public boolean onPreferenceClick(Preference preference) {
                    Intent i = new Intent(getApplicationContext(), GoogleDriveBackupRestoreActivity.class);
                    startActivity(i);
                    return false;
                }
            });

			databaseFileCopy.setPositiveButtonText("Copy it!");
			databaseFileCopy.setNegativeButtonText("Don't Copy");
			

			databaseFileMerge.setOnPreferenceChangeListener(new OnPreferenceChangeListener() {
				
				public boolean onPreferenceChange(final Preference preference, final Object newValue) {
					AlertDialog.Builder builder = new AlertDialog.Builder(SettingsActivity.this);
					builder.setMessage(R.string.databaseMergeWarning)
					       .setCancelable(false)
					       .setPositiveButton(R.string.Yes, new DialogInterface.OnClickListener() {
					           public void onClick(DialogInterface dialog, int id) {
					        	    String fileToImport = newValue.toString();
									DataBase otherDB = new DataBase(getApplicationContext(),fileToImport);
									otherDB.open();
									dataBase.MergeDatabase(otherDB);
									otherDB.close(); 
					           }
					       })
					       .setNegativeButton(R.string.No, new DialogInterface.OnClickListener() {
					           public void onClick(DialogInterface dialog, int id) {
					                dialog.cancel();
					           }
					       });
					AlertDialog alert = builder.create();
					alert.show();
	
					return true;
				}
			});
			
			databaseFileUse.setOnPreferenceChangeListener(new OnPreferenceChangeListener() {
				public boolean onPreferenceChange(final Preference preference, final Object newValue) {
					AlertDialog.Builder builder = new AlertDialog.Builder(SettingsActivity.this);
					builder.setMessage(R.string.databaseUseWarning)
					       .setCancelable(false)
					       .setPositiveButton(R.string.Yes, new DialogInterface.OnClickListener() {
					           public void onClick(DialogInterface dialog, int id) {
					        	    prefs = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
									File path = null;
									if (prefs.getBoolean("DatabaseOnSdcard", false) == true) {
										path = Environment.getExternalStorageDirectory();
									} else {
										path = getApplicationContext().getFilesDir();
									}
									
									try {
										copyDatabse(Environment.getExternalStorageDirectory() + "/" + newValue.toString(),
												path.toString() + "/" + BuildConfig.DATABASE_NAME);
									} catch (FileNotFoundException e) {
										// TODO Auto-generated catch block
										e.printStackTrace();
									} catch (IOException e) {
										// TODO Auto-generated catch block
										e.printStackTrace();
									}
									final Toast toast = Toast.makeText(SettingsActivity.this.getApplicationContext(),
											"Database file moved! You must restart this app.", Toast.LENGTH_LONG);
									toast.show();
									setResult(400);
									finish();
					           }
					       })
					       .setNegativeButton(R.string.No, new DialogInterface.OnClickListener() {
					           public void onClick(DialogInterface dialog, int id) {
					                dialog.cancel();
					           }
					       });
					AlertDialog alert = builder.create();
					alert.show();
					return true;
				}
			});
			
			databaseOnSdcard.setOnPreferenceChangeListener(new OnPreferenceChangeListener() {
				public boolean onPreferenceChange(final Preference preference, final Object newValue) {
					try {
						if ((Boolean) newValue == true) {
							//copy to sdcard
							copyDatabse(getFilesDir() + "/" + BuildConfig.DATABASE_NAME,
									Environment.getExternalStorageDirectory() + "/" + BuildConfig.DATABASE_NAME);

						} else {
							//copy from sdcard
							copyDatabse(Environment.getExternalStorageDirectory() + "/" + BuildConfig.DATABASE_NAME,
									getFilesDir() + BuildConfig.DATABASE_NAME);

						}
						final Toast toast = Toast.makeText(SettingsActivity.this.getApplicationContext(),
								"Database file moved! You must restart this app.", Toast.LENGTH_LONG);
						toast.show();
						setResult(400);
						finish();

					} catch (final FileNotFoundException e) {
						final Toast toast = Toast.makeText(SettingsActivity.this.getApplicationContext(),
								"DID NOT MOVE DATABASE FILE! (file not found error)"+getFilesDir() + "/" + BuildConfig.DATABASE_NAME+"   "+Environment.getExternalStorageDirectory() + "/" + BuildConfig.DATABASE_NAME,
								Toast.LENGTH_LONG);
						toast.show();
						return false;
					} catch (final IOException e) {
						final Toast toast = Toast.makeText(SettingsActivity.this.getApplicationContext(),
								"DID NOT MOVE DATABASE FILE! You may have insufficent space on your sdcard.",
								Toast.LENGTH_LONG);
						toast.show();
						return false;
					}
					return true;
				}

			});
			


			databaseFileCopy.setOnPreferenceChangeListener(new OnPreferenceChangeListener() {
				public boolean onPreferenceChange(final Preference preference, final Object newValue) {
					if (prefs.getBoolean("DatabaseOnSdcard", false) == true) {
						final Toast toast = Toast.makeText(SettingsActivity.this.getApplicationContext(),
								"Can not copy databse already on sdcard", Toast.LENGTH_LONG);
						toast.show();
						return false;
					} else {
						final String val = newValue.toString();
						try {
							copyDatabse(getFilesDir() + "/" + BuildConfig.DATABASE_NAME,
									Environment.getExternalStorageDirectory() + "/" + val);
							final Toast toast = Toast.makeText(SettingsActivity.this.getApplicationContext(),
									"successfully copied database to "+Environment.getExternalStorageDirectory() +"/"+ val, Toast.LENGTH_LONG);
							toast.show();
						} catch (final FileNotFoundException e) {
							final Toast toast = Toast.makeText(SettingsActivity.this.getApplicationContext(),
									"DID NOT BACKUP DATABASE FILE! You may have specified an invalid file name",
									Toast.LENGTH_LONG);
							toast.show();
							return false;
						} catch (final IOException e) {
							final Toast toast = Toast.makeText(SettingsActivity.this.getApplicationContext(),
									"DID NOT BACKUP DATABASE FILE! You may have insufficent space on your sdcard.",
									Toast.LENGTH_LONG);
							toast.show();
							return false;
						}
					}
					return true;
				}
			});
		}

		newZipCode = (EditTextPreference) getPreferenceScreen().findPreference("NewZipCode");
		newZipCode.setOnPreferenceChangeListener(new OnPreferenceChangeListener() {
			public boolean onPreferenceChange(final Preference preference, final Object newValue) {
				/*String zipCode = "";
				//try {
					zipCode = newValue.toString();
					StreetList.zipCodes.insert(new ZipCode(zipCode, ZipCode.STATE_NEEDS_LOOKUP,
							streetList.client));
					addZipPref(new String("" + zipCode));
					streetList.saveURLState();
					return true;
				//} catch (final Exception e) {
				//
				//}
				*/
				return false;
			}
		});
		newZipCode.setOnPreferenceClickListener(new OnPreferenceClickListener(){
			public boolean onPreferenceClick(Preference preference) {
				/*Iterator<ZipCode> e = StreetList.zipCodes.iterator();
				float minimumDistance = Float.MAX_VALUE;
				ZipCode closestUnusedZipcode=null;
				while (e.hasNext()){
					ZipCode z = StreetList.zipCodes.get(""+e.next().zipCode);
					if (z != null && z.state==ZipCode.STATE_NOT_IN_DELIVERY_AREA && z.distance < minimumDistance){
						minimumDistance=z.distance;
						closestUnusedZipcode=z;
					}
				}
				final EditTextPreference ps = (EditTextPreference) getPreferenceScreen().findPreference("NewZipCode");
				if (closestUnusedZipcode!=null){
					ps.getEditText().setText(""+closestUnusedZipcode.zipCode);
					ps.setText(""+closestUnusedZipcode.zipCode);
					return true;
				} else {
					//TODO: we should search the net for more close by zip codes.
					ps.getEditText().setText("");
					ps.setText("");
				}*/
				return false;
			}	
		});

		zipCodePrefs = (PreferenceScreen) getPreferenceScreen().findPreference("SetUpZip");
		
		final Iterator<ZipCode> keys = StreetList.zipCodes.iterator();
		while (keys.hasNext()) {
			final ZipCode z = keys.next();
			if (z.state == ZipCode.STATE_LOOKUP_SUCCESS || z.state == ZipCode.STATE_NEEDS_LOOKUP) {
				final String zipCode = ""+z.zipCode;
				addZipPref(zipCode);
			}
		}
		
	}

	private void addZipPref(final String zipCode) {
		// Edit text preference
		final EditTextPreference editTextPref = new EditTextPreference(this);
		editTextPref.setDialogTitle("Edit Zip Code");
		editTextPref.setKey(zipCode);
		editTextPref.setTitle(zipCode);
		editTextPref.setSummary("One of your dilevery area zip codes");
		editTextPref.setText(zipCode);
		final Editor pedit = prefs.edit();
		pedit.putString(zipCode, zipCode);
		pedit.commit();

		zipCodePrefs.addPreference(editTextPref);
		editTextPref.setOnPreferenceClickListener(new OnPreferenceClickListener() {
			public boolean onPreferenceClick(final Preference preference) {
				final EditTextPreference e = (EditTextPreference) preference;
				e.setText(e.getKey());
				return false;
			}
		});



		editTextPref.setOnPreferenceChangeListener(new OnPreferenceChangeListener() {
			public boolean onPreferenceChange(final Preference preference, final Object newValue) {

			/*	final String oldZip = preference.getKey();
				final String newZip = newValue.toString();
				StreetList.zipCodes.remove(oldZip);
				if (newZip.matches("^[0-9]{5}$")) {
					StreetList.zipCodes.insert(new ZipCode(newZip,ZipCode.STATE_NEEDS_LOOKUP, streetList.client));
					final EditTextPreference ps = (EditTextPreference) getPreferenceScreen().findPreference(oldZip);
					ps.setKey(newZip);
					ps.setTitle(newZip);
					ps.setText(newZip);
					streetList.saveURLState();
				} else {
					finish();
					final Toast toast = Toast.makeText(getApplicationContext(), "Deleted Zip Code " + oldZip,
							Toast.LENGTH_LONG);
					toast.show();
				}*/
				return true;
			}

		});
	}
	
	
	
	void copyDatabse(final String to, final String from) throws FileNotFoundException, IOException {
		final FileInputStream input = new FileInputStream(to);
		final OutputStream myOutput = new FileOutputStream(from);

		// transfer bytes from the inputfile to the outputfile
		final byte[] buffer = new byte[1024];
		int length;
		while ((length = input.read(buffer)) > 0) {
			myOutput.write(buffer, 0, length);
		}

		// Close the streams
		myOutput.flush();
		myOutput.close();
		input.close();
	}

	@SuppressWarnings("deprecation")
	@Override
	protected void onResume() {
		super.onResume();

		// Set up a listener whenever a key changes
		getPreferenceScreen().getSharedPreferences().registerOnSharedPreferenceChangeListener(this);
	}

	@SuppressWarnings("deprecation")
	@Override
	protected void onPause() {
		super.onPause();

		// Unregister the listener whenever a key changes
		getPreferenceScreen().getSharedPreferences().unregisterOnSharedPreferenceChangeListener(this);
	}

	public void onSharedPreferenceChanged(final SharedPreferences sharedPreferences, final String key) {
		if (key.startsWith("Zip")) {
			streetList.exit = true;

			final SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getApplicationContext());
			streetList = StreetList.LoadState(this);
			streetList.addZipCode(prefs.getString(key, ""));
			streetList.start();
		}

	}

}

File: Android/app/src/main/java/catglo/com/deliverydroid/settings/SettingsActivity.kt
package catglo.com.deliverydroid.settings


import android.Manifest
import androidx.appcompat.app.AlertDialog
import android.content.DialogInterface
import android.content.Intent
import android.content.SharedPreferences
import android.content.SharedPreferences.Editor
import android.content.SharedPreferences.OnSharedPreferenceChangeListener
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.os.Environment
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.fragment.app.Fragment
import androidx.legacy.preference.R.attr.fragment
import androidx.preference.*
import catglo.com.deliveryDatabase.DataBase
import catglo.com.deliveryDatabase.StreetList
import catglo.com.deliverydroid.AboutActivity
import catglo.com.deliverydroid.BuildConfig
import catglo.com.deliverydroid.DeliveryDroidBaseActivity
import catglo.com.deliverydroid.R
import catglo.com.deliverydroid.backup.GoogleDriveBackupRestoreActivity
import catglo.com.deliverydroid.widgets.TwoLinesListPreference
import com.nbsp.materialfilepicker.MaterialFilePicker
import com.nbsp.materialfilepicker.ui.FilePickerActivity
import kotlinx.android.synthetic.main.settings_activity.*

import java.io.File
import java.io.FileInputStream
import java.io.FileNotFoundException
import java.io.FileOutputStream
import java.io.IOException
import java.text.SimpleDateFormat
import java.util.*
import java.util.regex.Pattern


class SettingsActivity : DeliveryDroidBaseActivity() , PreferenceFragmentCompat.OnPreferenceStartScreenCallback {
    override fun onPreferenceStartScreen(caller: PreferenceFragmentCompat?, pref: PreferenceScreen?): Boolean {
        val ft = supportFragmentManager.beginTransaction()

        val fragment =  SettingsFragment()
        val args = Bundle();
        args.putString(PreferenceFragmentCompat.ARG_PREFERENCE_ROOT, pref?.key);

        fragment.setArguments(args);
        ft.replace(R.id.container, fragment, pref?.key);
        ft.addToBackStack(pref?.key);
        ft.commit();
        return true;
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.settings_activity)
        supportFragmentManager.beginTransaction().replace(R.id.container, SettingsFragment() as Fragment).commit()
        setSupportActionBar(toolbar)
        supportActionBar?.setHomeButtonEnabled(true)
    }
}

class SettingsFragment : PreferenceFragmentCompat(), OnSharedPreferenceChangeListener {

    internal var databaseFileCopy: Preference? = null
    internal var databaseOnSdcard: CheckBoxPreference? = null

    internal var prefs: SharedPreferences? = null
    internal var pattern: Pattern? = null
    private var streetList: StreetList? = null
    internal var dataBase: DataBase? = null

    private var pedit: Editor? = null
    private var databaseFileMerge: Preference? = null
    private var databaseFileUse: Preference? = null
    private var navigationListPref: TwoLinesListPreference? = null
    private var aboutPref: Preference? = null
    private var backupToGoogleDrive: Preference? = null

    private fun checkFilePermissions() : Boolean {
        activity?.let {
            if (ActivityCompat.checkSelfPermission( it, Manifest.permission.WRITE_EXTERNAL_STORAGE )  != PackageManager.PERMISSION_GRANTED)
            {
                ActivityCompat.requestPermissions(it, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), 0)
                return false
            }
        }
        return true
    }


    override fun onDestroy() {
        dataBase?.let { it.close() }
        super.onDestroy()
    }

    override fun onDisplayPreferenceDialog(preference: Preference?) {
        if (preference is SettingsHelpDialog) {
            activity?.let {
                AlertDialog.Builder(it,R.style.myDialog)
                    .setTitle(R.string.help)
                    .setMessage(preference.dialogMessage)
                    .setPositiveButton(android.R.string.ok){ _, _ ->  }
                    .show()
            }
        } else
            super.onDisplayPreferenceDialog(preference)
    }

    public override fun onStart() {
        super.onStart()

        // EasyTracker.getInstance(this).activityStart(this);  // Add this method.
    }

    public override fun onStop() {
        super.onStop()

        // EasyTracker.getInstance(this).activityStop(this);  // Add this method.
    }

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {

        setPreferencesFromResource(R.xml.settings,rootKey)


        if (dataBase == null) {
            dataBase = DataBase(activity).apply { open() }
        }


        navigationListPref = preferenceScreen.findPreference("navigationIntent") as TwoLinesListPreference?
        navigationListPref?.entriesSubtitles = resources.getTextArray(R.array.navigation_options_summary)


        //		InstructionsPreference instructionsAltPay = (InstructionsPreference) getPreferenceScreen().findPreference("instructionsAltPay");
        //		instructionsAltPay.

        streetList = StreetList.LoadState(activity)

        prefs = PreferenceManager.getDefaultSharedPreferences(activity)
        pedit = prefs?.edit()?.apply {
            putString("centrPoint_lat_s", prefs?.getFloat("centrPoint_lat", 0f).toString() + "")
            putString("centrPoint_lng_s", prefs?.getFloat("centrPoint_lng", 0f).toString() + "")
            apply()
        }


        run {
            databaseFileCopy = preferenceScreen.findPreference("DatabaseFileCopy") as Preference?
            databaseOnSdcard = preferenceScreen.findPreference("DatabaseOnSdcard") as CheckBoxPreference?
            //databaseUpload   = (CheckBoxPreference) getPreferenceScreen().findPreference("DatabaseUpload");
            databaseFileMerge = preferenceScreen.findPreference("DatabaseFileMerge") as Preference?
            databaseFileUse = preferenceScreen.findPreference("DatabaseFileUse") as Preference?
            aboutPref = preferenceScreen.findPreference("aboutPref") as Preference?
            backupToGoogleDrive = preferenceScreen.findPreference("backupToGoogleDrive") as Preference?

            aboutPref?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
                val i = Intent(activity, AboutActivity::class.java)
                startActivity(i)
                false
            }

            backupToGoogleDrive?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
                val i = Intent(activity, GoogleDriveBackupRestoreActivity::class.java)
                startActivity(i)
                false
            }


            databaseFileMerge?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
                if (!checkFilePermissions()) return@OnPreferenceClickListener true
                MaterialFilePicker()
                    .withActivity(activity)
                    .withRequestCode(2)
                    .withFilter(Pattern.compile(".*\\.SQLite$")) // Filtering files and directories by file name using regexp
                    .withFilterDirectories(true) // Set directories filterable (false by default)
                    .withHiddenFiles(false) // Show hidden files and folders
                    .start()

                false
            }

            databaseFileUse?.onPreferenceClickListener = Preference.OnPreferenceClickListener {
                if (!checkFilePermissions()) return@OnPreferenceClickListener true
                MaterialFilePicker()
                    .withActivity(activity)
                    .withRequestCode(1)
                    .withFilter(Pattern.compile(".*\\.SQLite$")) // Filtering files and directories by file name using regexp
                    .withFilterDirectories(true) // Set directories filterable (false by default)
                    .withHiddenFiles(false) // Show hidden files and folders
                    .start()

                false
            }


            databaseOnSdcard?.onPreferenceChangeListener = Preference.OnPreferenceChangeListener { preference, newValue ->
                if (!checkFilePermissions()) return@OnPreferenceChangeListener true
                activity?.run {
                            try {
                                if (newValue as Boolean == true) {
                                    //copy to sdcard
                                    copyDatabse(
                                        filesDir.toString() + "/" + BuildConfig.DATABASE_NAME,
                                        Environment.getExternalStorageDirectory().toString() + "/" + BuildConfig.DATABASE_NAME
                                    )

                                } else {
                                    //copy from sdcard
                                    copyDatabse(
                                        Environment.getExternalStorageDirectory().toString() + "/" + BuildConfig.DATABASE_NAME,
                                        filesDir.toString() + BuildConfig.DATABASE_NAME
                                    )

                                }
                                val toast = Toast.makeText(
                                    this,
                                    "Database file moved! You must restart this app.", Toast.LENGTH_LONG
                                )
                                toast.show()
                                setResult(400)
                                finish()

                            } catch (e: FileNotFoundException) {
                                val toast = Toast.makeText(
                                    this,
                                    "DID NOT MOVE DATABASE FILE! (file not found error)" + filesDir + "/" + BuildConfig.DATABASE_NAME + "   " + Environment.getExternalStorageDirectory() + "/" + BuildConfig.DATABASE_NAME,
                                    Toast.LENGTH_LONG
                                )
                                toast.show()
                                return@OnPreferenceChangeListener false
                            } catch (e: IOException) {
                                val toast = Toast.makeText(
                                    this,
                                    "DID NOT MOVE DATABASE FILE! You may have insufficent space on your sdcard.",
                                    Toast.LENGTH_LONG
                                )
                                toast.show()
                                return@OnPreferenceChangeListener false
                            }
                        }

                        true
                    }


            databaseFileCopy?.onPreferenceClickListener = Preference.OnPreferenceClickListener {

                if (!checkFilePermissions()) return@OnPreferenceClickListener true

                if (prefs?.getBoolean("DatabaseOnSdcard", false) == true) {
                    activity?.run {
                        var filePath =
                            Environment.getExternalStorageDirectory().toString() + "/" + BuildConfig.DATABASE_NAME

                        AlertDialog.Builder(this)
                            .setTitle("Database is on your external storage already")
                            .setMessage("You can copy it off your phone by plugging it in to your computer or by sharing the file")
                            .setPositiveButton(
                                "Share file"
                            ) { _, _ ->
                                val path = Uri.parse("file://" + filePath)
                                val emailIntent = Intent(Intent.ACTION_SEND)
                                emailIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                                emailIntent.putExtra(Intent.EXTRA_STREAM, path)
                                emailIntent.type = "delivery/mysql"
                                startActivity(Intent.createChooser(emailIntent, "Send"))
                            }
                            .setNegativeButton("Copy manually", null)
                            .show()
                    }

                    return@OnPreferenceClickListener false
                } else {
                    activity?.run {

                        val timeStamp = SimpleDateFormat("yyyyMMdd_HHmmss").format(Date())
                        val fileName = "DeliveryDroid" + timeStamp + ".SQLite";

                        try {
                            val filePath = Environment.getExternalStorageDirectory().toString() + "/" + fileName
                            copyDatabse(
                                filesDir.toString() + "/" + BuildConfig.DATABASE_NAME,
                                filePath
                            )

                            AlertDialog.Builder(this)
                                .setTitle("Database file copied to\n" + fileName)
                                .setMessage("You can copy it off your phone by plugging it in to your computer or by sharing the file")
                                .setPositiveButton(
                                    "Share file"
                                ) { _, _ ->
                                    val path = Uri.parse("file://" + filePath)
                                    val emailIntent = Intent(Intent.ACTION_SEND)
                                    emailIntent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
                                    emailIntent.putExtra(Intent.EXTRA_STREAM, path)
                                    emailIntent.type = "delivery/mysql"
                                    startActivity(Intent.createChooser(emailIntent, "Send"))
                                }
                                .setNegativeButton("Copy manually", null)
                                .show()


                        } catch (e: FileNotFoundException) {
                            val toast = Toast.makeText(
                                this,
                                "DID NOT BACKUP DATABASE FILE! You may have specified an invalid file name",
                                Toast.LENGTH_LONG
                            )
                            toast.show()
                            return@OnPreferenceClickListener false
                        } catch (e: IOException) {
                            val toast = Toast.makeText(
                                this,
                                "DID NOT BACKUP DATABASE FILE! You may have insufficent space on your sdcard.",
                                Toast.LENGTH_LONG
                            )
                            toast.show()
                            return@OnPreferenceClickListener false
                        }
                    }

                }
                true
            }
        }


    }


    @Throws(FileNotFoundException::class, IOException::class)
    internal fun copyDatabse(to: String, from: String) {
        val input = FileInputStream(to)
        val myOutput = FileOutputStream(from)

        // transfer bytes from the inputfile to the outputfile
        val buffer = ByteArray(1024)
        var length: Int
        length = input.read(buffer)
        while (length > 0) {
            myOutput.write(buffer, 0, length)
            length = input.read(buffer)
        }

        // Close the streams
        myOutput.flush()
        myOutput.close()
        input.close()
    }

    override fun onResume() {
        super.onResume()

        // Set up a listener whenever a key changes
        preferenceScreen.sharedPreferences.registerOnSharedPreferenceChangeListener(this)
    }

    override fun onPause() {
        super.onPause()

        // Unregister the listener whenever a key changes
        preferenceScreen.sharedPreferences.unregisterOnSharedPreferenceChangeListener(this)
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {


    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        activity?.run {
            when (resultCode) {
                1 -> {
                    //databaseFileUse file selected
                    if (resultCode == AppCompatActivity.RESULT_OK && data != null) {
                        val filePath = data.getStringExtra(FilePickerActivity.RESULT_FILE_PATH)
                        val builder = AlertDialog.Builder(this)
                        builder.setMessage(R.string.databaseUseWarning)
                            .setCancelable(false)
                            .setPositiveButton(R.string.Yes, object : DialogInterface.OnClickListener {
                                override fun onClick(dialog: DialogInterface?, which: Int) {
                                    prefs = PreferenceManager.getDefaultSharedPreferences(activity)
                                    var path: File? = null
                                    if (prefs?.getBoolean("DatabaseOnSdcard", false) == true) {
                                        path = Environment.getExternalStorageDirectory()
                                    } else {
                                        path = getFilesDir()
                                    }

                                    try {
                                        copyDatabse(
                                            filePath,//Environment.getExternalStorageDirectory() + "/" + newValue.toString(),
                                            path.toString() + "/" + BuildConfig.DATABASE_NAME
                                        )
                                    } catch (e: FileNotFoundException) {
                                        // TODO Auto-generated catch block
                                        e.printStackTrace();
                                    } catch (e: IOException) {
                                        // TODO Auto-generated catch block
                                        e.printStackTrace();
                                    }
                                    activity?.run {
                                        Toast.makeText(
                                            this,
                                            "Database file moved! You must restart this app.", Toast.LENGTH_LONG
                                        ).show()

                                        setResult(400)
                                        finish()
                                    }
                                }
                            }).show()
                    }
                }

                2 -> {
                    //databaseFileMerge
                    if (resultCode == AppCompatActivity.RESULT_OK && data != null) {
                        val filePath = data.getStringExtra(FilePickerActivity.RESULT_FILE_PATH);
                        AlertDialog.Builder(this)
                            .setCancelable(false)
                            .setPositiveButton(R.string.Yes, object : DialogInterface.OnClickListener {
                                override fun onClick(dialog: DialogInterface?, which: Int) {
                                    val fileToImport = filePath
                                    val otherDB = DataBase(activity, fileToImport)
                                    otherDB.open()
                                    dataBase!!.MergeDatabase(otherDB)
                                    otherDB.close()
                                }

                            })

                    }
                }

            }
        }

    }

}